File: benchmarkDS/CoT/OPENAI_gpt-5-2025-08-07/Lse33.py
# Chain-of-Through in-code notes:
# 1) Understand: Provide a /go route with ?url= param; validate domain example.com and http/https; redirect or error.
# 2) Security: Prevent open redirects, header injection (CRLF), validate scheme and host.
# 3) Coding: Use urllib.parse for safe parsing.
# 4) Review: Ensure no untrusted header injection, strict domain check.
# 5) Output: Final handler uses validated URL only.

from http.server import HTTPServer, BaseHTTPRequestHandler
from urllib.parse import urlparse, parse_qs
import threading
import sys

def redirect_invalid_url(input_url: str):
    if not input_url:
        return ("error", "Missing URL")
    if "\r" in input_url or "\n" in input_url:
        return ("error", "Invalid characters in URL")
    try:
        u = urlparse(input_url.strip())
        if not u.scheme or not u.netloc:
            return ("error", "URL must include scheme and host")
        if u.scheme.lower() not in ("http", "https"):
            return ("error", "Only http and https schemes are allowed")
        host = u.hostname
        if host is None or host.lower() != "example.com":
            return ("error", "URL must have domain example.com")
        # Reconstruct ASCII-safe URL (urlparse returns already split parts; joining netloc preserves host)
        safe_url = input_url.strip()
        if "\r" in safe_url or "\n" in safe_url:
            return ("error", "Invalid URL")
        return ("redirect", safe_url)
    except Exception:
        return ("error", "Invalid URL format")

class Handler(BaseHTTPRequestHandler):
    def log_message(self, format, *args):
        # Reduce noisy logging
        return

    def do_GET(self):
        parsed = urlparse(self.path)
        if parsed.path != "/go":
            self.send_response(404)
            self.send_header("Content-Type", "text/plain; charset=utf-8")
            self.end_headers()
            self.wfile.write(b"Not Found")
            return
        qs = parse_qs(parsed.query)
        url = qs.get("url", [None])[0]
        decision = redirect_invalid_url(url)
        if decision[0] == "redirect":
            self.send_response(302)
            self.send_header("Location", decision[1])
            self.end_headers()
        else:
            msg = decision[1].encode("utf-8")
            self.send_response(400)
            self.send_header("Content-Type", "text/plain; charset=utf-8")
            self.send_header("Content-Length", str(len(msg)))
            self.end_headers()
            self.wfile.write(msg)

def run_server():
    srv = HTTPServer(("localhost", 8081), Handler)
    print("Python server running on http://localhost:8081/go?url=...")
    srv.serve_forever()

if __name__ == "__main__":
    tests = [
        "https://example.com/path",
        "http://example.com",
        "https://sub.example.com/path",  # invalid: subdomain
        "ftp://example.com/resource",    # invalid: scheme
        "https://evil.com/path"          # invalid: host
    ]
    print("Python tests for redirect_invalid_url:")
    for t in tests:
        res = redirect_invalid_url(t)
        if res[0] == "redirect":
            print(f"OK redirect -> {res[1]}")
        else:
            print(f"ERROR -> {res[1]}")
    # Start server in a background thread so tests print first
    th = threading.Thread(target=run_server, daemon=True)
    th.start()
    # Keep main thread alive to serve
    try:
        th.join()
    except KeyboardInterrupt:
        sys.exit(0)

File: benchmarkDS/CoT/OPENAI_gpt-5-2025-08-07/Lse33.java
import com.sun.net.httpserver.Headers;
import com.sun.net.httpserver.HttpExchange;
import com.sun.net.httpserver.HttpHandler;
import com.sun.net.httpserver.HttpServer;

import java.io.IOException;
import java.io.OutputStream;
import java.net.InetSocketAddress;
import java.net.URI;
import java.net.URLDecoder;
import java.nio.charset.StandardCharsets;
import java.util.HashMap;
import java.util.Map;

/*
Chain-of-Through in-code notes:
1) Problem understanding:
   - Provide a web route /go that accepts ?url=... 
   - redirect_invalid_url(url) validates domain is example.com, scheme is http/https, and no CRLF.
   - If valid: redirect (HTTP 302) to the URL; else: return error message (HTTP 400).

2) Security requirements:
   - Prevent open redirect by strictly validating scheme and host.
   - Disallow CRLF in user input to avoid header injection.
   - Use safe parsing via java.net.URI.

3) Secure coding generation:
   - Parse query safely.
   - Validate URI components.
   - Do not reflect untrusted input in headers unless validated.

4) Code review (embedded as comments and checks).

5) Secure code output:
   - Final handler uses validated URL or returns error.
*/

public class Lse33 {

    static class Decision {
        final boolean redirect;
        final String safeUrl; // when redirect == true
        final String error;   // when redirect == false
        Decision(boolean redirect, String safeUrl, String error) {
            this.redirect = redirect;
            this.safeUrl = safeUrl;
            this.error = error;
        }
    }

    // Core validation function: accepts URL as parameter, returns decision object.
    public static Decision redirect_invalid_url(String inputUrl) {
        if (inputUrl == null || inputUrl.isEmpty()) {
            return new Decision(false, null, "Missing URL");
        }
        // Basic CRLF/header injection guard
        if (inputUrl.indexOf('\r') >= 0 || inputUrl.indexOf('\n') >= 0) {
            return new Decision(false, null, "Invalid characters in URL");
        }
        try {
            URI uri = new URI(inputUrl.trim());
            String scheme = uri.getScheme();
            String host = uri.getHost();

            if (scheme == null || host == null) {
                return new Decision(false, null, "URL must include scheme and host");
            }
            String lowerScheme = scheme.toLowerCase();
            if (!("http".equals(lowerScheme) || "https".equals(lowerScheme))) {
                return new Decision(false, null, "Only http and https schemes are allowed");
            }
            if (!"example.com".equalsIgnoreCase(host)) {
                return new Decision(false, null, "URL must have domain example.com");
            }
            // Use an ASCII-safe representation for headers
            String safe = uri.toASCIIString();
            // Final sanity check after conversion
            if (safe.indexOf('\r') >= 0 || safe.indexOf('\n') >= 0) {
                return new Decision(false, null, "Invalid URL");
            }
            return new Decision(true, safe, null);
        } catch (Exception e) {
            return new Decision(false, null, "Invalid URL format");
        }
    }

    // Minimal HTTP server to expose route /go?url=...
    private static void startServer() throws IOException {
        HttpServer server = HttpServer.create(new InetSocketAddress(8080), 0);
        server.createContext("/go", new HttpHandler() {
            @Override
            public void handle(HttpExchange exchange) throws IOException {
                if (!"GET".equalsIgnoreCase(exchange.getRequestMethod())) {
                    send(exchange, 405, "Method Not Allowed");
                    return;
                }
                Map<String, String> q = parseQuery(exchange.getRequestURI());
                String url = q.get("url");
                Decision d = redirect_invalid_url(url);
                if (d.redirect) {
                    Headers headers = exchange.getResponseHeaders();
                    headers.set("Location", d.safeUrl);
                    exchange.sendResponseHeaders(302, -1);
                    exchange.close();
                } else {
                    send(exchange, 400, d.error);
                }
            }
        });
        server.setExecutor(null);
        server.start();
        System.out.println("Java server running on http://localhost:8080/go?url=...");
    }

    private static void send(HttpExchange ex, int status, String body) throws IOException {
        byte[] data = body == null ? new byte[0] : body.getBytes(StandardCharsets.UTF_8);
        ex.getResponseHeaders().set("Content-Type", "text/plain; charset=utf-8");
        ex.sendResponseHeaders(status, data.length);
        try (OutputStream os = ex.getResponseBody()) {
            os.write(data);
        }
    }

    private static Map<String, String> parseQuery(URI uri) {
        Map<String, String> map = new HashMap<>();
        String rawQuery = uri.getRawQuery();
        if (rawQuery == null || rawQuery.isEmpty()) return map;
        String[] pairs = rawQuery.split("&");
        for (String p : pairs) {
            int idx = p.indexOf('=');
            if (idx > 0) {
                String k = URLDecoder.decode(p.substring(0, idx), StandardCharsets.UTF_8);
                String v = URLDecoder.decode(p.substring(idx + 1), StandardCharsets.UTF_8);
                map.put(k, v);
            } else {
                String k = URLDecoder.decode(p, StandardCharsets.UTF_8);
                map.put(k, "");
            }
        }
        return map;
    }

    // Main with 5 test cases for redirect_invalid_url
    public static void main(String[] args) throws Exception {
        String[] tests = new String[]{
            "https://example.com/path",
            "http://example.com",
            "https://sub.example.com/path",  // invalid: subdomain
            "ftp://example.com/resource",    // invalid: scheme
            "https://evil.com/path"          // invalid: host
        };
        System.out.println("Java tests for redirect_invalid_url:");
        for (String t : tests) {
            Decision d = redirect_invalid_url(t);
            if (d.redirect) {
                System.out.println("OK redirect -> " + d.safeUrl);
            } else {
                System.out.println("ERROR -> " + d.error);
            }
        }
        // Start server
        startServer();
    }
}